src/control/system/raft.go

```diff
@@ -84,6 +92,151 @@ func (ro raftOp) String() string {
+		return svc.LeadershipTransfer().Error()
+	}); err != nil {
+		return errors.Wrap(err, cause.Error())
+	}
+	return cause
+}
+
+// ShutdownRaft signals that the raft implementation should shut down
+// and release any resources it is holding. Blocks until the shutdown
+// is complete.
+func (db *Database) ShutdownRaft() error {
+	db.log.Debug("shutting down raft instance")
+	return db.raft.withReadLock(func(svc raftService) error {
+		return svc.Shutdown().Error()
+	})
+}
+
+// configureRaft sets up the raft service in preparation for starting it.
+func (db *Database) configureRaft() error {
+	if db.raftTransport == nil {
+		return errors.New("no raft transport configured")
+	}
+
+	rc := raft.DefaultConfig()
+	rc.Logger = newHcLogger(db.log)
+	rc.SnapshotThreshold = 16 // arbitrarily low to exercise snapshots
+	//rc.SnapshotInterval = 5 * time.Second
+	rc.HeartbeatTimeout = 250 * time.Millisecond
+	rc.ElectionTimeout = 250 * time.Millisecond
+	rc.LeaderLeaseTimeout = 125 * time.Millisecond
+	rc.LocalID = raft.ServerID(db.serverAddress())
+
+	snaps, err := raft.NewFileSnapshotStoreWithLogger(db.cfg.RaftDir, 2, rc.Logger)
+	if err != nil {
+		return err
+	}
+
+	sysDBPath := filepath.Join(db.cfg.RaftDir, sysDBFile)
+	boltDB, err := boltdb.NewBoltStore(sysDBPath)
+	if err != nil {
+		return err
+	}
+
+	r, err := raft.NewRaft(rc, (*fsm)(db), boltDB, boltDB, snaps, db.raftTransport)
+	if err != nil {
+		return err
+	}
+	db.raft.setSvc(r)
+
+	return nil
+}
+
+// startRaft is responsible for configuring and starting the raft service
+// on this node. If shouldBootstrap is true, then the service will be started
+// in a special bootstrap mode that does not require a quorum.
+func (db *Database) startRaft(shouldBootstrap bool) error {
+	db.log.Debugf("isBootstrap: %t, shouldBootstrap: %t", db.IsBootstrap(), shouldBootstrap)
+
+	if db.IsBootstrap() && shouldBootstrap {
+		// Rank 0 is reserved for the first instance on the bootstrap server.
+		db.data.NextRank = 1
+
+		db.log.Debugf("bootstrapping MS on %s", db.replicaAddr)
+		bsc := raft.Configuration{
+			Servers: []raft.Server{
+				{
+					Suffrage: raft.Voter,
+					ID:       raft.ServerID(db.getReplica().String()),
+					Address:  raft.ServerAddress(db.getReplica().String()),
+				},
+			},
+		}
+		if err := db.raft.withReadLock(func(svc raftService) error {
+			if f := svc.BootstrapCluster(bsc); f.Error() != nil {
+				return errors.Wrapf(f.Error(), "failed to bootstrap raft instance on %s", db.getReplica())
+			}
+			return nil
+		}); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// loggingTransport is used to wrap a raft.Transport with methods that
+// log the action and arguments. Used for debugging while the feature is
+// being developed.
+type loggingTransport struct {
+	raft.Transport
+	log logging.Logger
+}
+
+/*
+func (dt *loggingTransport) AppendEntriesPipeline(id raft.ServerID, target raft.ServerAddress) (raft.AppendPipeline, error) {
```

Comment:
TODO to give information about re-enabling?

Reply:
This is debug scaffolding, will be removed in the final hardening phase of this effort. The commented-out implementations are sometimes useful for local debugging but way too spammy to leave enabled otherwise.